### PR TITLE
  ALSA: hda/hdmi: fix failures at PCM open on Intel ICL and later

### PR DESCRIPTION
**Pull request by kv2019i**
ALSA: hda/hdmi: fix failures at PCM open on Intel ICL and later

Bugfix and one trace improvement patch.

Fixes: #1978
Fixes: #2216
Fixes: #2217

### CODE REVIEW

File: sound/pci/hda/patch_hdmi.c [hunk 1]
```diff
@@ -1804,33 +1804,43 @@ static int hdmi_add_cvt(struct hda_codec *codec, hda_nid_t cvt_nid)
...
 		if (!(caps & AC_WCAP_DIGITAL))
 			continue;
 
-		switch (type) {
-		case AC_WID_AUD_OUT:
-			hdmi_add_cvt(codec, nid);
-			break;
-		case AC_WID_PIN:
+		if (get_wcaps_type(caps) == AC_WID_AUD_OUT)
+			hdmi_add_cvt(codec, start_nid + i);
```
Comment (ranj063) [inline sound/pci/hda/patch_hdmi.c]: @kv2019i just hdmi_add_cvt(codec, nid); should be enough here?

Comment (kv2019i) [reply sound/pci/hda/patch_hdmi.c]: Aa, thanks @ranj063 , forgot to to replace that when adding the local nid variable. Will fix.

**Comment by plbossart** [comment]
@kv2019i I'll let you merge this, i don't recall if this is the final version or still WIP.

**Comment by libinyang** [comment]
@kv2019i Good to catch the initialization issue. 

If we go through the code deeply, it seems we should not set intel_hsw_fixup flag for ICL as the latest platforms are not hsw like platform. Without the flag, the connection list can be initialized correctly. We are wrongly use this flag now. And I can also see some potential issues if we set intel_hsw_fixup flag on ICL.

**Comment by kv2019i** [comment]
@libinyang wrote:
> If we go through the code deeply, it seems we should not set intel_hsw_fixup flag for ICL as the latest platforms are not hsw 
> like platform. Without the flag, the connection list can be initialized correctly. We are wrongly use this flag now. And I can also > see some potential issues if we set intel_hsw_fixup flag on ICL.

That's a good point, thanks. Let me do some tests on this. The hsw_fixup flag is used for multiple things and I don't think all have been fixed for ICL. E.g. it's used for hdmi_add_pin(). But if the snd_hda_get_raw_connections() works correctly on ICL, we could indeed remote the exception for ICL and newer.

@plbossart @ranj063 Let's not merge this yet.

**Comment by kv2019i** [comment]
@libinyang OK, I tried without the hsw_fixup, but it still fails. On pins with monitor connected, snd_hda_get_raw_connections() works ok and full list of cvt nids is returned, but on pins with no monitor connected at time of probe, no connections are returned.

So maybe best to go with current patch still.

UPDATE: @libinyang I tested with ICL and CFL systems without the fixup and the results are similar, so I'll keep the fixup.

**Comment by kv2019i** [comment]
I sent this upstream now, let's see if we need to adjust more before merging to sof-dev.

**Comment by kv2019i** [comment]
Merged upstream. As this fixes multiple SOF issues, I'll go ahead and merge the upstream version to sof-dev right away.

**Pull request merged**